Add Dhyana6060BSI support

KuktaeKim opened the pull request:
In this pull request, I’ve added several new features and improvements:

1. Added Tucsen_CXP_Dhyana6060BSI.xml
This file provides support for the Tucsen Dhyana6060BSI (AXIS-SXRF-60) camera.

2. Added support for info autosaveFields to makeDb.py

3. Added ADGenICam::readFloat64()
This method is necessary to allow reading of floating-point fields such as DeviceTemperature and SensorTemperature of the camera using SCAN fields.
Note: To support on-demand reading of additional registers, other data types' asyn read functions will need to be implemented in the future.

4. Set PINI=YES on output records
This change addresses an issue where the camera resets its registers to factory defaults after a restart, causing mismatches with IOC autosaved values. Setting PINI=YES ensures that the autosaved values are correctly pushed back to the camera during IOC startup.

5. Added call to stopCapture() when ROI, Offset, or Binning is modified during acquisition
This prevents IOC crashes that could occur when changing image size, offset, or binning while images are being captured.

KuktaeKim committed:
Add Tucsen_CXP_Dhyana6060BSI.xml for the Tucsen Dhyana6060BSI (AXIS-SXRF-60) Camera

MarkRivers commented:
I have serious concerns about setting PINI=YES for all output records (bo, ao, int64out, mbbo, stringout).
The problem is that there are many GenICam features that have more than one output record to control them.  There is a generic output record which is part of ADCore, for example TriggerMode, TriggerSource, MinX, and SizeY.  There are also camera-specific output records to control these same features, generate in makeDb.py.  For these features those records are called GC_TriggerMode, GC_TriggerSource, GC_OffsetY, and GC_Height.

This is a screenshot of my ADSpinnaker screen.  It shows that TriggerMode=On, TriggerSource=Line0, MinY=400, and SizeY=1190.
![image](https://github.com/user-attachments/assets/eca40984-69c9-4804-b072-52cf2d21996b)

This is a screenshot of the first GenICam screen for this detector.  It shows that GC_TriggerMode=Off and GC_TriggerSource=Software.  Note that input (_RBV) records agree with the ADSpinnaker screen, but the output records do not agree.
![image](https://github.com/user-attachments/assets/2c00f48f-6494-46ed-9d56-2c4d51650c80)

This is a screenshot of the second GenICam screen for this detector.  It shows that GC_OffsetY=0 and GC_Height=2048.  Note that input (_RBV) records agree with the ADSpinnaker screen, but the output records do not agree.
![image](https://github.com/user-attachments/assets/5d6829d7-08b3-4488-8197-02809d07bd2c)

The problem is that there if we accept your PR there are now 2 output records with PINI=YES for the same GenICam feature and they have conflicting values.  We cannot predict which one will be processed last, so we cannot predict what the actual state of the camera will be when the IOC restarts.

This is a serious issue.  Have you actually tested this PR and looked for this effect?

KuktaeKim commented:
Hi @MarkRivers,

Thank you for your insightful review.

At the beginning of the IOC development for this camera, we noticed that several PVs had the same functionality but different names and did not match, such as MinX and GC_OffsetX, MinY and GC_OffsetY, SizeX and GC_Width, SizeY and GC_Height, etc. To address this, we initially created some PVs to match these values. After that, based on user feedback, we set PINI=YES, and we haven’t encountered any issues so far.

However, if you believe this could lead to a serious issue, I’m happy to close this PR and create a new one without the PINI=YES change.

henriquesimoes reviewed (changes requested):
Besides the issue Mark mentioned, I have other concerns. Most of them I commented quoting the changes themselves below.

Regarding the commits, I find that af0fc9ee ("Run latest makeDb.py from Mark Rivers.") commit message is misleading. It mentions `makeDb.py`, but it updates ADL files. If you just wanted to update the files because you ran `makeAdl.py` with an older version, I'd rather you squash af0fc9ee with 0d4c67ee. Alternatively, one commit for each script you run.

> However, if you believe this could lead to a serious issue, I’m happy to close this PR and create a new one without the PINI=YES change.

If we agree on keeping some changes you proposed, you don't need to close and open another PR. You can simply [`git-rebase(1)`](https://git-scm.com/docs/git-rebase)  your changes (to drop commits, to squash them, fix them, change their messages, etc), and force push your new commits in this same branch (and thus this PR). This way, we keep the review comments in a single place, and the final patch series with reasonable commit messages and correct changes.

If the proposal changes significantly, then I agree it makes sense to close and open another request.

MarkRivers commented on scripts/makeDb.py:
```diff
@@ -145,6 +145,13 @@ def is_node_readonly(node):
             ro = True
     return ro
 
+# Which fields to autosave for different record types
+a_autosaveFields        = 'DESC LOLO LOW HIGH HIHI LLSV LSV HSV HHSV EGU TSE PREC'
+b_autosaveFields        = 'DESC ZSV OSV TSE'
+long_autosaveFields     = 'DESC LOLO LOW HIGH HIHI LLSV LSV HSV HHSV EGU TSE'
+mbb_autosaveFields      = 'DESC ZRSV ONSV TWSV THSV FRSV FVSV SXSV SVSV EISV NISV TESV ELSV TVSV TTSV FTSV FFSV TSE'
+string_autosaveFields   = 'DESC TSE'
```
I agree that it should be removed.

MarkRivers commented on GenICamApp/src/ADGenICam.cpp:
```diff
@@ -78,6 +81,25 @@ ADGenICam::ADGenICam(const char *portName, size_t maxMemory, int priority, int s
   * Takes action if the function code requires it.  ADAcquire, ADSizeX, and many other
   * function codes make calls to the underlying library from this function. */
 
+asynStatus ADGenICam::readFloat64( asynUser *pasynUser, epicsFloat64 *value)
+{
+    asynStatus status = asynSuccess;
+    int function = pasynUser->reason;
+    static const char *functionName = "readFloat64";
+
+    GenICamFeature *pFeature = mGCFeatureSet.getByIndex(function);
+    if (pFeature) {
+        pFeature->read(value, true);
+    }
+
+    asynPrint(pasynUserSelf, ASYN_TRACEIO_DRIVER,
+        "%s::%s function=%d, value=%f, status=%d\n",
+        driverName, functionName, function, *value, status);
+
+    callParamCallbacks();
+    return status;
+}
+
```
@bhill-slac i agree. For consistency I think that readInt32() and readOctet() should be implemented in a similar way.

KuktaeKim commented:
> Besides the issue Mark mentioned, I have other concerns. Most of them I commented quoting the changes themselves below.
> 
> Regarding the commits, I find that [af0fc9e](https://github.com/areaDetector/ADGenICam/commit/af0fc9ee6e8def25fdb0666e4fbed669161ae5d9) ("Run latest makeDb.py from Mark Rivers.") commit message is misleading. It mentions `makeDb.py`, but it updates ADL files. If you just wanted to update the files because you ran `makeAdl.py` with an older version, I'd rather you squash [af0fc9e](https://github.com/areaDetector/ADGenICam/commit/af0fc9ee6e8def25fdb0666e4fbed669161ae5d9) with [0d4c67e](https://github.com/areaDetector/ADGenICam/commit/0d4c67eede8d3c05d3cadaa3fb2ed727db326b04). Alternatively, one commit for each script you run.
> 
> > However, if you believe this could lead to a serious issue, I’m happy to close this PR and create a new one without the PINI=YES change.
> 
> If we agree on keeping some changes you proposed, you don't need to close and open another PR. You can simply [`git-rebase(1)`](https://git-scm.com/docs/git-rebase) your changes (to drop commits, to squash them, fix them, change their messages, etc), and force push your new commits in this same branch (and thus this PR). This way, we keep the review comments in a single place, and the final patch series with reasonable commit messages and correct changes.
> 
> If the proposal changes significantly, then I agree it makes sense to close and open another request.

@henriquesimoes I squashed https://github.com/areaDetector/ADGenICam/commit/af0fc9ee6e8def25fdb0666e4fbed669161ae5d9 with https://github.com/areaDetector/ADGenICam/commit/0d4c67eede8d3c05d3cadaa3fb2ed727db326b04. 
Thank you for your recommendation.

KuktaeKim commented on GenICamApp/src/ADGenICam.cpp:
```diff
@@ -139,6 +139,10 @@ asynStatus ADGenICam::writeInt32( asynUser *pasynUser, epicsInt32 value)
              (function == ADMinY)        ||
              (function == ADBinX)        ||
              (function == ADBinY)) {
+		if (acquiring) {
+			status = stopCapture();
+			asynPrint(pasynUserSelf, ASYN_TRACEIO_DRIVER, "%s%s: called stopCapture() for ROI change\n", driverName, functionName);
+		}
```
Thank you for the advice. We will create a subclass in the ADEureys.

KuktaeKim commented:
I squashed [commit af0fc9e](https://github.com/areaDetector/ADGenICam/commit/af0fc9ee6e8def25fdb0666e4fbed669161ae5d9) with [commit 0d4c67e](https://github.com/areaDetector/ADGenICam/commit/0d4c67eede8d3c05d3cadaa3fb2ed727db326b04). I also removed all other commits unrelated to the Tucsen Dhyana6060BSI (AXIS-SXRF-60) camera support.

I’ll submit separate pull requests for the other changes.

henriquesimoes reviewed (commented):
@KuktaeKim, thanks for reworking the PR.

To review your automated changes, I tried regenerating the database template generation on top of e55044d9 based on your XML file expecting to see no diff, but I got the following difference in the resulting file.

<details>
<summary>Resulting diff</summary>

```diff
diff --git a/GenICamApp/Db/Tucsen_CXP_Dhyana6060BSI.template b/GenICamApp/Db/Tucsen_CXP_Dhyana6060BSI.template
index 985f9b4..c553985 100644
--- a/GenICamApp/Db/Tucsen_CXP_Dhyana6060BSI.template
+++ b/GenICamApp/Db/Tucsen_CXP_Dhyana6060BSI.template
@@ -47,6 +47,12 @@ record(stringin, "$(P)$(R)GC_DeviceUserID_RBV") {
   field(DISA, "0")
 }
 
+record(stringout, "$(P)$(R)GC_DeviceUserID") {
+  field(DTYP, "asynOctetWrite")
+  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_S_DeviceUserID")
+  field(DISA, "0")
+}
+
 record(mbbi, "$(P)$(R)GC_DeviceScanType_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_DeviceScanType")
@@ -58,17 +64,6 @@ record(mbbi, "$(P)$(R)GC_DeviceScanType_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_DeviceScanType") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_DeviceScanType")
-  field(DOL,  "0")
-  field(ZRST, "Areascan")
-  field(ZRVL, "0")
-  field(ONST, "Linescan")
-  field(ONVL, "1")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_DeviceInitStatus_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_DeviceInitStatus")
@@ -76,12 +71,6 @@ record(ai, "$(P)$(R)GC_DeviceInitStatus_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_DeviceInitStatus") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_DeviceInitStatus")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_DeviceWorkingTime_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_DeviceWorkingTime")
@@ -89,12 +78,6 @@ record(ai, "$(P)$(R)GC_DeviceWorkingTime_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_DeviceWorkingTime") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_DeviceWorkingTime")
-  field(DISA, "0")
-}
-
 record(longout, "$(P)$(R)GC_DeviceReset") {
   field(DTYP, "asynInt32")
   field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_C_DeviceReset")
@@ -131,13 +114,6 @@ record(ai, "$(P)$(R)GC_DeviceTemperature_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_DeviceTemperature") {
-  field(DTYP, "asynFloat64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_DeviceTemperature")
-  field(PREC, "3")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_DevWarTemperature_RBV") {
   field(DTYP, "asynFloat64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_DeviceWarningTemperature")
@@ -146,13 +122,6 @@ record(ai, "$(P)$(R)GC_DevWarTemperature_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_DevWarTemperature") {
-  field(DTYP, "asynFloat64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_DeviceWarningTemperature")
-  field(PREC, "3")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_SensorTemperature_RBV") {
   field(DTYP, "asynFloat64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_SensorTemperature")
@@ -161,13 +130,6 @@ record(ai, "$(P)$(R)GC_SensorTemperature_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_SensorTemperature") {
-  field(DTYP, "asynFloat64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_SensorTemperature")
-  field(PREC, "3")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_SetSenTemperature_RBV") {
   field(DTYP, "asynFloat64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_SetSensorTemperature")
@@ -321,12 +283,6 @@ record(ai, "$(P)$(R)GC_SensorWidth_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_SensorWidth") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_SensorWidth")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_SensorHeight_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_SensorHeight")
@@ -334,12 +290,6 @@ record(ai, "$(P)$(R)GC_SensorHeight_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_SensorHeight") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_SensorHeight")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_SensorPixelSize_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorPixelSize")
@@ -349,15 +299,6 @@ record(mbbi, "$(P)$(R)GC_SensorPixelSize_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_SensorPixelSize") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorPixelSize")
-  field(DOL,  "0")
-  field(ZRST, "10umx10um")
-  field(ZRVL, "0")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_SensorShutterType_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorShutterType")
@@ -367,15 +308,6 @@ record(mbbi, "$(P)$(R)GC_SensorShutterType_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_SensorShutterType") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorShutterType")
-  field(DOL,  "0")
-  field(ZRST, "Rolling")
-  field(ZRVL, "0")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_GainMode_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GainMode")
@@ -459,17 +391,6 @@ record(mbbi, "$(P)$(R)GC_SensorADWidth_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_SensorADWidth") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorADWidth")
-  field(DOL,  "12")
-  field(ZRST, "12bit")
-  field(ZRVL, "12")
-  field(ONST, "14bit")
-  field(ONVL, "14")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_SensorPGAHighGain_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorPGAHighGain")
@@ -485,21 +406,6 @@ record(mbbi, "$(P)$(R)GC_SensorPGAHighGain_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_SensorPGAHighGain") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorPGAHighGain")
-  field(DOL,  "0")
-  field(ZRST, "G0p8x")
-  field(ZRVL, "0")
-  field(ONST, "G2p0x")
-  field(ONVL, "1")
-  field(TWST, "G6p0x")
-  field(TWVL, "2")
-  field(THST, "G6p6x")
-  field(THVL, "3")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_SensorPGALowGain_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorPGALowGain")
@@ -515,21 +421,6 @@ record(mbbi, "$(P)$(R)GC_SensorPGALowGain_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_SensorPGALowGain") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_SensorPGALowGain")
-  field(DOL,  "0")
-  field(ZRST, "G0p8x")
-  field(ZRVL, "0")
-  field(ONST, "G2p0x")
-  field(ONVL, "1")
-  field(TWST, "G6p0x")
-  field(TWVL, "2")
-  field(THST, "G6p6x")
-  field(THVL, "3")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_PixelFormat_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_PixelFormat")
@@ -543,19 +434,6 @@ record(mbbi, "$(P)$(R)GC_PixelFormat_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_PixelFormat") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_PixelFormat")
-  field(DOL,  "0x00000103")
-  field(ZRST, "Mono12")
-  field(ZRVL, "0x00000103")
-  field(ONST, "Mono14")
-  field(ONVL, "0x00000104")
-  field(TWST, "Mono16")
-  field(TWVL, "0x00000105")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_Width_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Width")
@@ -1072,19 +950,6 @@ record(mbbi, "$(P)$(R)GC_GpsStatus_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_GpsStatus") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GpsStatus")
-  field(DOL,  "0")
-  field(ZRST, "UnLocked")
-  field(ZRVL, "0")
-  field(ONST, "Locked")
-  field(ONVL, "1")
-  field(TWST, "NoGps")
-  field(TWVL, "2")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_GpsLatitudeRef_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GpsLatitudeRef")
@@ -1096,17 +961,6 @@ record(mbbi, "$(P)$(R)GC_GpsLatitudeRef_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_GpsLatitudeRef") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GpsLatitudeRef")
-  field(DOL,  "0")
-  field(ZRST, "North")
-  field(ZRVL, "0")
-  field(ONST, "South")
-  field(ONVL, "1")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_GpsLatitude_RBV") {
   field(DTYP, "asynFloat64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_GpsLatitude")
@@ -1115,13 +969,6 @@ record(ai, "$(P)$(R)GC_GpsLatitude_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_GpsLatitude") {
-  field(DTYP, "asynFloat64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_GpsLatitude")
-  field(PREC, "3")
-  field(DISA, "0")
-}
-
 record(mbbi, "$(P)$(R)GC_GpsLongitudeRef_RBV") {
   field(DTYP, "asynInt32")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GpsLongitudeRef")
@@ -1133,17 +980,6 @@ record(mbbi, "$(P)$(R)GC_GpsLongitudeRef_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_GpsLongitudeRef") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GpsLongitudeRef")
-  field(DOL,  "0")
-  field(ZRST, "East")
-  field(ZRVL, "0")
-  field(ONST, "West")
-  field(ONVL, "1")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_GpsLongitude_RBV") {
   field(DTYP, "asynFloat64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_GpsLongitude")
@@ -1152,13 +988,6 @@ record(ai, "$(P)$(R)GC_GpsLongitude_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_GpsLongitude") {
-  field(DTYP, "asynFloat64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_D_GpsLongitude")
-  field(PREC, "3")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_GpsDate_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GpsDate")
@@ -1166,12 +995,6 @@ record(ai, "$(P)$(R)GC_GpsDate_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_GpsDate") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GpsDate")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_GpsTime_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GpsTime")
@@ -1179,12 +1002,6 @@ record(ai, "$(P)$(R)GC_GpsTime_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_GpsTime") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GpsTime")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_GpsTrigStartTime_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GpsTrigStartTime")
@@ -1305,25 +1122,6 @@ record(mbbi, "$(P)$(R)GC_GenerateStatus_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_GenerateStatus") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_GenerateStatus")
-  field(DOL,  "0")
-  field(ZRST, "GenFree")
-  field(ZRVL, "0")
-  field(ONST, "GenBusy")
-  field(ONVL, "1")
-  field(TWST, "CalBusy")
-  field(TWVL, "2")
-  field(THST, "WRBusy")
-  field(THVL, "3")
-  field(FRST, "GenDone")
-  field(FRVL, "4")
-  field(FVST, "GenStop")
-  field(FVVL, "5")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_GenerateGrayValue_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GenerateGrayValue")
@@ -1331,12 +1129,6 @@ record(ai, "$(P)$(R)GC_GenerateGrayValue_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_GenerateGrayValue") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_GenerateGrayValue")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_Standard_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Standard")
@@ -1344,12 +1136,6 @@ record(ai, "$(P)$(R)GC_Standard_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_Standard") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Standard")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_Revision_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Revision")
@@ -1357,12 +1143,6 @@ record(ai, "$(P)$(R)GC_Revision_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_Revision") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Revision")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_XmlManifestSize_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlManifestSize")
@@ -1370,12 +1150,6 @@ record(ai, "$(P)$(R)GC_XmlManifestSize_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_XmlManifestSize") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlManifestSize")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_XmlManSelector_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlManifestSelector")
@@ -1396,12 +1170,6 @@ record(ai, "$(P)$(R)GC_XmlVersion_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_XmlVersion") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlVersion")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_XmlSchemaVersion_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlSchemaVersion")
@@ -1409,12 +1177,6 @@ record(ai, "$(P)$(R)GC_XmlSchemaVersion_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_XmlSchemaVersion") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlSchemaVersion")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_XmlUrlAddress_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlUrlAddress")
@@ -1422,12 +1184,6 @@ record(ai, "$(P)$(R)GC_XmlUrlAddress_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_XmlUrlAddress") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_XmlUrlAddress")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_DeviceLinkID_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_DeviceLinkID")
@@ -1435,12 +1191,6 @@ record(ai, "$(P)$(R)GC_DeviceLinkID_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_DeviceLinkID") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_DeviceLinkID")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_MasterHostLinkID_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_MasterHostLinkID")
@@ -1461,12 +1211,6 @@ record(ai, "$(P)$(R)GC_ConPacketDataSize_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_ConPacketDataSize") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_ControlPacketDataSize")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_StrPacketDataSize_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_StreamPacketDataSize")
@@ -1591,43 +1335,6 @@ record(mbbi, "$(P)$(R)GC_ConConfigDefault_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_ConConfigDefault") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_ConnectionConfigDefault")
-  field(DOL,  "0x00010028")
-  field(ZRST, "x1_CXP_1")
-  field(ZRVL, "0x00010028")
-  field(ONST, "x1_CXP_2")
-  field(ONVL, "0x00010030")
-  field(TWST, "x1_CXP_3")
-  field(TWVL, "0x00010038")
-  field(THST, "x1_CXP_5")
-  field(THVL, "0x00010040")
-  field(FRST, "x1_CXP_6")
-  field(FRVL, "0x00010048")
-  field(FVST, "x2_CXP_1")
-  field(FVVL, "0x00020028")
-  field(SXST, "x2_CXP_2")
-  field(SXVL, "0x00020030")
-  field(SVST, "x2_CXP_3")
-  field(SVVL, "0x00020038")
-  field(EIST, "x2_CXP_5")
-  field(EIVL, "0x00020040")
-  field(NIST, "x2_CXP_6")
-  field(NIVL, "0x00020048")
-  field(TEST, "x4_CXP_1")
-  field(TEVL, "0x00040028")
-  field(ELST, "x4_CXP_2")
-  field(ELVL, "0x00040030")
-  field(TVST, "x4_CXP_3")
-  field(TVVL, "0x00040038")
-  field(TTST, "x4_CXP_5")
-  field(TTVL, "0x00040040")
-  field(FTST, "x4_CXP_6")
-  field(FTVL, "0x00040048")
-  field(DISA, "0")
-}
-
 record(longout, "$(P)$(R)GC_ConnectionReset") {
   field(DTYP, "asynInt32")
   field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_C_ConnectionReset")
@@ -1734,15 +1441,6 @@ record(mbbi, "$(P)$(R)GC_TapGeometry_RBV") {
   field(DISA, "0")
 }
 
-record(mbbo, "$(P)$(R)GC_TapGeometry") {
-  field(DTYP, "asynInt32")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_E_TapGeometry")
-  field(DOL,  "0")
-  field(ZRST, "X1_1Y")
-  field(ZRVL, "0")
-  field(DISA, "0")
-}
-
 record(ai, "$(P)$(R)GC_Image1StreamID_RBV") {
   field(DTYP, "asynInt64")
   field(INP,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Image1StreamID")
@@ -1750,9 +1448,3 @@ record(ai, "$(P)$(R)GC_Image1StreamID_RBV") {
   field(DISA, "0")
 }
 
-record(ao, "$(P)$(R)GC_Image1StreamID") {
-  field(DTYP, "asynInt64")
-  field(OUT,  "@asyn($(PORT),$(ADDR=0),$(TIMEOUT=1))GC_I_Image1StreamID")
-  field(DISA, "0")
-}
-
```

</details>

Here goes the command I ran:

```shell
git checkout e55044d9
python3 scripts/makeDb.py xml/Tucsen_CXP_Dhyana6060BSI.xml GenICamApp/Db/Tucsen_CXP_Dhyana6060BSI.template
```

Which version of the ADGenICam did you use to run `makeDb.py` script in your commit?

It seems to have generated a bunch of disabled records which aren't created in the latest version of `makeDb.py`.

henriquesimoes commented:
Hi, @KuktaeKim.

> Please take another look when you have a chance.

Sorry for the long delay, I came back from my vacation this week. I repeated my test, and saw the same result. It looks like [your latest push](https://github.com/areaDetector/ADGenICam/compare/e55044d929b646e2dd8dca1751a0b451c140e744..c99996755c8dd3e3e76485ff8d0bf5e52da4e7a8) only included changes to ADL files.

Maybe you didn't push the latest changes? :thinking:

MarkRivers commented:
@KuktaeKim I suggest just committing the XML file, or e-mail it to me.  I can make the .template and .adl files, plus all of the autoconvert files to Phoebus, edm, etc.

KuktaeKim commented:
@henriquesimoes @MarkRivers  Sorry for repeatedly making mistakes and bothering you, and thank you for your recommendation. I’ve updated the PR to commit only the XML file for the Tucsen Dhyana6060BSI (AXIS-SXRF-60) camera.

MarkRivers commented:
I have merged this PR.   I pulled it to my system and created the .template and OPI files.  I pushed these back to Github.

henriquesimoes commented:
> Sorry for repeatedly making mistakes and bothering you, and thank you for your recommendation.

Not a bother at all. I'm happy to review and help getting the change set in a good shape. Thank you for your contribution.

> I have merged this PR. I pulled it to my system and created the .template and OPI files. I pushed these back to Github.

Thanks.
